dcicutils/qa_utils.py

```diff
@@ -551,8 +559,150 @@ def download_fileobj(self, Bucket, Key, Fileobj, **kwargs):  # noqa - Uppercase
+    # but will return the object if it's in the instance.
+    value = getattr(thing, key)
+    if isinstance(value, CustomizableProperty):
+        # This is an uncustomized instance variable, not a class variable.
+        # That's not an intended use case, but just report it without involving mention of the class.
+        raise UncustomizedInstance(instance=thing, field=key)
+    else:
+        return value
+
+
+class VersionChecker:
+
+    PYPROJECT = CustomizableProperty('PYPROJECT', description="The repository-relative name of the pyproject file.")
+    CHANGELOG = CustomizableProperty('CHANGELOG', description="The repository-relative name of the change log.")
+
+    @classmethod
+    def check_version(cls):
+        version = cls._check_version()
+        if getattr_customized(cls, "CHANGELOG"):
+            cls._check_change_history(version)
+
+    @classmethod
+    def _check_version(cls):
+
+        __tracebackhide__ = True
+
+        pyproject_file = getattr_customized(cls, 'PYPROJECT')
+        assert os.path.exists(pyproject_file), "Missing pyproject file: %s" % pyproject_file
+        pyproject = toml.load(pyproject_file)
+        version = pyproject.get('tool', {}).get('poetry', {}).get('version', None)
+        assert version, "Missing version in %s." % pyproject_file
+        PRINT("Version = %s" % version)
+        return version
+
+    VERSION_LINE_PATTERN = re.compile("^[#* ]*([0-9]+[.][^ \t\n]*)([ \t\n].*)?$")
+
+    @classmethod
+    def _check_change_history(cls, version=None):
+
+        changelog_file = getattr_customized(cls, "CHANGELOG")
+
+        if not changelog_file:
+            if version:
+                raise AssertionError("Cannot check version without declaring a CHANGELOG file.")
+            return
+
+        assert os.path.exists(changelog_file), "Missing changelog file: %s" % changelog_file
+
+        with io.open(changelog_file) as fp:
+            versions = []
+            for line in fp:
+                m = cls.VERSION_LINE_PATTERN.match(line)
+                if m:
+                    versions.append(m.group(1))
+
+        assert versions, "No version info was parsed from %s" % changelog_file
+        # Might be sorted top to bottom or bottom to top, but ultimately the current version should be first or last.
+        assert versions[0] == version or versions[-1] == version, (
+                "Missing entry for version %s in %s." % (version, changelog_file)
+        )
+
+
+def raises_regexp(error_class, pattern):
+    # Mostly compatible with unittest style, so that (approximately):
+    #  pytest.raises(error_class, match=exp) == unittest.TestCase.assertRaisesRegexp(error_class, exp)
+    # They differ on what to do if an error_class other than the expected one is raised.
+    return pytest.raises(error_class, match=pattern)
+
+
+def check_duplicated_items_by_key(key, items, url=None, formatter=str):
+
+    __tracebackhide__ = True
```

Comment:
In PyTest it causes the the function with this in it not to be the focus of the error message. Instead its caller is treated as the one that had the error, as if this function were internal to 'raise'. Our `check_true` function should have this, for example. I just ran across it when trying to write `raises_regexp` but then ended up solving this a different way because it turned out `pytest` had the functionality built in (and implemented using this feature).